test/e2e/aws_test.go

```diff
@@ -118,6 +140,172 @@ var _ = Describe("functional tests", func() {
 	})
 })
 
+func scaleMachineDeployment(namespace, machineDeployment string, replicasCurrent int32, replicasProposed int32) {
+	fmt.Fprintf(GinkgoWriter, "Scaling MachineDeployment from %d to %d\n", replicasCurrent, replicasProposed)
+	deployment := &clusterv1.MachineDeployment{}
+	Expect(kindClient.Get(context.TODO(), apimachinerytypes.NamespacedName{Namespace: namespace, Name: machineDeployment}, deployment))
+	deployment.Spec.Replicas = &replicasProposed
+	Expect(kindClient.Update(context.TODO(), deployment)).NotTo(HaveOccurred())
+	waitForMachinesCountMatch(namespace, machineDeployment, replicasCurrent, replicasProposed)
+	waitForMachineDeploymentRunning(namespace, machineDeployment, replicasProposed)
+}
+
+func createMachineTemplate(namespace, clusterName, machineDeploymentName, awsMachineTemplateName, bootstrapConfigName string,nodepool string, replicas int32) {
+	fmt.Fprintf(GinkgoWriter, "Creating MachineDeployment in namespace %s with name %s\n", namespace, machineDeploymentName)
+	makeAWSMachineTemplate(namespace, awsMachineTemplateName)
+	makeJoinBootstrapConfigTemplate(namespace, bootstrapConfigName)
+	makeMachineDeployment(namespace, machineDeploymentName, awsMachineTemplateName, bootstrapConfigName, clusterName, nodepool, replicas)
+	waitForMachinesCountMatch(namespace, machineDeploymentName,replicas, replicas)
+	waitForMachineDeploymentRunning(namespace, machineDeploymentName, replicas)
```

Comment:
It might be okay to just skip waitForMachinesCountMatch and just use waitForMachineDeploymentRunning

Reply:
n case of scaling down the MachineDeployment, Replicas and ReadyReplicas will be set to replica value immediately after marking the machines for deletion. But the actual deletion will be happening in the back ground. So, the waitForMachineDeploymentRunning() method call will get executed quickly and the next operation of scale UP/DOWN will be delayed as the previous tasks were not completed. Hence, i wanted to ensure that scale down is completed in terms of machines.

> I set the milestone to v0.4.x given that the PR is targeting the 0.4 release branch, any reason why not target master instead?

master branch is having v1alpha3 changes which were causing tests to fail. After checking with @detiber , raised it against release-0.4